### ISSUE TITLE
`MLP` scaling w.r.t. `expansion_factor`

### ISSUE THREAD

**Issue by EIFY** [graphcore-research/unit-scaling#70]
Hi, is the gradient scaling of `MLP` correct w.r.t. `expansion_factor`?

I noticed that in [demo.ipynb](https://github.com/graphcore-research/unit-scaling/blob/50ea961f1c0f2939a4bd7c8b9ada41acf9a583e4/examples/demo.ipynb), the output of `show_layer_stats(UmupTransformerLayer(128, 0), (batch_size, sequence_length, 128))` is
```python
# UmupTransformerLayer:
              output.std = 1.01
          input.grad.std = 1.10
     attn_qkv.weight.std = 1.00
     attn_out.weight.std = 1.00
       mlp_up.weight.std = 1.00
     mlp_gate.weight.std = 1.00
     mlp_down.weight.std = 1.00
attn_qkv.weight.grad.std = 0.64
attn_out.weight.grad.std = 1.09
  mlp_up.weight.grad.std = 0.71
mlp_gate.weight.grad.std = 0.73
mlp_down.weight.grad.std = 1.00
```
among which `mlp_up.weight.grad.std` and `mlp_gate.weight.grad.std` are quite far from unity. I reproduced the results and traced it down to the minimal module that behaves the same, which is `uu.MLP` itself:
```python
batch_size = 16
show_layer_stats(uu.MLP(hidden_size=128, expansion_factor), (batch_size, 128))
```
```python
# MLP (expansion_factor=1):
(...)
linear_1.weight.grad.std = 0.94
linear_gate.weight.grad.std = 0.99
(...)
# MLP (expansion_factor=2):
(...)
linear_1.weight.grad.std = 0.73
linear_gate.weight.grad.std = 0.72
(...)
# MLP (expansion_factor=4):
(...)
linear_1.weight.grad.std = 0.51
linear_gate.weight.grad.std = 0.53
(...)
# MLP (expansion_factor=8):
(...)
linear_1.weight.grad.std = 0.35
linear_gate.weight.grad.std = 0.37
(...)
# MLP (expansion_factor=16):
(...)
linear_1.weight.grad.std = 0.25
linear_gate.weight.grad.std = 0.26
(...)
# MLP (expansion_factor=64):
(...)
linear_1.weight.grad.std = 0.13
linear_gate.weight.grad.std = 0.13
(...)
```
So experimentally both `linear_1.weight.grad.std` and `linear_gate.weight.grad.std` are proportional to `1/sqrt(expansion_factor)`. This dependency goes away if I scale the gradient accordingly:
```python
import unit_scaling.scale as S
import math

class MLP(nn.Module):
    def __init__(
        self,
        hidden_size: int,
        expansion_factor: int = 4,
        constraint: Optional[str] = "to_output_scale",
    ) -> None:
        super().__init__()
        self.expansion_factor = expansion_factor
        intermediate_size = hidden_size * expansion_factor
        self.linear_1 = uu.Linear(hidden_size, intermediate_size, constraint=constraint)
        self.linear_gate = uu.Linear(hidden_size, intermediate_size, constraint=constraint)
        self.linear_2 = uu.Linear(intermediate_size, hidden_size, constraint=constraint)

    def forward(self, input: Tensor) -> Tensor:
        input = S.scale_bwd(input, 1 / math.sqrt(self.expansion_factor))
        z = U.silu_glu(self.linear_1(input), self.linear_gate(input))
        z = S.scale_bwd(z, math.sqrt(self.expansion_factor))
        return self.linear_2(z)
```
`show_layer_stats(MLP(hidden_size=128, expansion_factor=64), (batch_size, 128))`:
```python
# MLP:
              output.std = 1.04
          input.grad.std = 1.51
     linear_1.weight.std = 1.00
  linear_gate.weight.std = 1.00
     linear_2.weight.std = 1.00
linear_1.weight.grad.std = 1.07
linear_gate.weight.grad.std = 1.09
linear_2.weight.grad.std = 1.06
```
Less importantly I also noticed that `uu.MLP`'s `input.grad.std` is closer to `1.5` and `uu.MHSA`'s `linear_qkv.weight.grad.std` is closer to `0.6` when tested. Both are O(1) but they do propagate if there is no norm layer:
```python
class UmupTransformerLayer(nn.Module):
    def __init__(self, width: int, layer_idx: int, norm=uu.LayerNorm) -> None:
        super().__init__()
        self.attn_norm = norm(width)
        # (...)
        self.mlp_norm = norm(width)
        # (...)
```
```python
# UmupTransformerLayer (default):
              output.std = 1.01
          input.grad.std = 1.10
     attn_qkv.weight.std = 1.00
     attn_out.weight.std = 1.00
       mlp_up.weight.std = 1.00
     mlp_gate.weight.std = 1.00
     mlp_down.weight.std = 1.00
attn_qkv.weight.grad.std = 0.63
attn_out.weight.grad.std = 1.09
  mlp_up.weight.grad.std = 0.71
mlp_gate.weight.grad.std = 0.73
mlp_down.weight.grad.std = 1.00
```
```python
# UmupTransformerLayer (norm=nn.Identity):
              output.std = 1.06
          input.grad.std = 1.20
     attn_qkv.weight.std = 1.00
     attn_out.weight.std = 1.00
       mlp_up.weight.std = 1.00
     mlp_gate.weight.std = 1.00
     mlp_down.weight.std = 1.00
attn_qkv.weight.grad.std = 0.89
attn_out.weight.grad.std = 1.53
  mlp_up.weight.grad.std = 0.91
mlp_gate.weight.grad.std = 0.94
mlp_down.weight.grad.std = 1.30
```

**Comment by DouglasOrr** [comment]
Hi @EIFY, thanks for your careful work investigating this - very helpful! First, my apologies for being very slow replying to your message - I was waiting to discuss with colleagues, but should have just replied. Sorry about that. Given the delay, you may have worked this all out already, but just in case...

## So experimentally both linear_1.weight.grad.std and linear_gate.weight.grad.std are proportional to 1/sqrt(expansion_factor).

You're right, as confirmed by your experiments. The reason for this is that unit scaling uses local scaling rules, which must have constraints applied in order to keep the gradients consistent with the forward pass.

For a linear layer, we first compute the ideal scales ([U.linear](https://github.com/graphcore-research/unit-scaling/blob/50ea961f1c0f2939a4bd7c8b9ada41acf9a583e4/unit_scaling/functional.py#L223)):

```python
output_scale = 1 / fan_in ** 0.5
grad_input_scale = 1 / fan_out ** 0.5
grad_weight_scale = grad_bias_scale = 1 / batch_size ** 0.5
```

Then we constrain them, if required, so that we never add two gradients with different ratios of scale_fwd/scale_bwd (this is somewhat explained in the [unit scaling paper](https://arxiv.org/abs/2303.11257), section 5.1, but the [u-mup paper](https://arxiv.org/abs/2407.17465), Appendix H, is perhaps more accessible).

```python
output_scale, grad_input_scale = apply_constraint(
    constraint, output_scale, grad_input_scale
)
```

It's this step which introduces the mis-scaling. The default constraint for `uu.MLP` is `"to_output_scale"`, which means that the `output_scale` is used for both `output_scale` and `grad_input_scale`. This means that the gradients `z.grad` (in [uu.MLP](https://github.com/graphcore-research/unit-scaling/blob/50ea961f1c0f2939a4bd7c8b9ada41acf9a583e4/unit_scaling/_modules.py#L455)) are under-scaled when `expansion_factor > 1`. This then means that `linear_1` and `linear_gate` parameter gradients are under-scaled. This can be fixed with "non-local knowledge", where we would allow separate scales for forward and backward passes in the MLP, because we know that these scales will cancel each other out before any tensors are added. However, this can be risky, so the unit-scaling library takes the more conservative approach of constraining the scales.

To check this, we can turn it off:

```python
show_layer_stats(uu.MLP(512, expansion_factor=64, constraint=None), (2, 512))
# MLP:
              output.std = 1.05
          input.grad.std = 1.46
     linear_1.weight.std = 1.00
  linear_gate.weight.std = 1.00
     linear_2.weight.std = 1.00
linear_1.weight.grad.std = 1.05
linear_gate.weight.grad.std = 1.10
linear_2.weight.grad.std = 1.08
```

I hadn't considered before, but I think this is safe in this case of an MLP, since both up-projection's `scale_fwd` are the same and equal to the down-projection's `scale_bwd`; this makes me wonder if it should be the default setting for an MLP.

## Less importantly I also noticed that uu.MLP's input.grad.std is closer to 1.5 and uu.MHSA's linear_qkv.weight.grad.std is closer to 0.6 when tested. Both are O(1) but they do propagate if there is no norm layer.

Agreed - I believe that this is again due to the constraints (some of which are not disabled with `constraint=None`), this time in `U.silu_glu` in the MLP, and `scaled_dot_product_attention` in MHSA. Unfortunately it isn't safe to disable these constraints, as the residual layer would cause incorrect gradients. In this case, the unit_scaling library has to choose between a scale-preserving forwards pass, and a scale-preserving backwards pass, and the rule is to use "to_output_scale" to preserve scale in the forwards pass.

To check this, I tried a regular MLP with no gate & GLU layer, and the input scale was approximately 1.

---

I hope this helps - I will discuss with my colleague r.e. updating the default constraint for the MLP, but please feel free to ask any further questions. Thanks!
